// @antv/f6-hammerjs/lib/touchactionjs/touchaction-Consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TOUCH_ACTION_MAP = exports.TOUCH_ACTION_PAN_Y = exports.TOUCH_ACTION_PAN_X = exports.TOUCH_ACTION_NONE = exports.TOUCH_ACTION_MANIPULATION = exports.TOUCH_ACTION_COMPUTE = exports.TOUCH_ACTION_AUTO = void 0;

var _utilsConsts = require("../utils/utils-consts");

var _getTouchactionProps = _interopRequireDefault(require("./get-touchaction-props"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// magical touchAction value
const TOUCH_ACTION_COMPUTE = 'compute';
exports.TOUCH_ACTION_COMPUTE = TOUCH_ACTION_COMPUTE;
const TOUCH_ACTION_AUTO = 'auto';
exports.TOUCH_ACTION_AUTO = TOUCH_ACTION_AUTO;
const TOUCH_ACTION_MANIPULATION = 'manipulation'; // not implemented

exports.TOUCH_ACTION_MANIPULATION = TOUCH_ACTION_MANIPULATION;
const TOUCH_ACTION_NONE = 'none';
exports.TOUCH_ACTION_NONE = TOUCH_ACTION_NONE;
const TOUCH_ACTION_PAN_X = 'pan-x';
exports.TOUCH_ACTION_PAN_X = TOUCH_ACTION_PAN_X;
const TOUCH_ACTION_PAN_Y = 'pan-y';
exports.TOUCH_ACTION_PAN_Y = TOUCH_ACTION_PAN_Y;
const TOUCH_ACTION_MAP = (0, _getTouchactionProps.default)();
exports.TOUCH_ACTION_MAP = TOUCH_ACTION_MAP;
