// @antv/f2/lib/components/point/withPoint.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));
var _jsx = require("../../jsx");
var _geometry = _interopRequireDefault(require("../geometry"));
var _default = function _default(View) {
  return /*#__PURE__*/function (_Geometry) {
    (0, _inherits2.default)(Point, _Geometry);
    var _super = (0, _createSuper2.default)(Point);
    function Point() {
      (0, _classCallCheck2.default)(this, Point);
      return _super.apply(this, arguments);
    }
    (0, _createClass2.default)(Point, [{
      key: "getDefaultCfg",
      value: function getDefaultCfg() {
        return {
          geomType: 'point'
        };
      }
    }, {
      key: "render",
      value: function render() {
        var props = this.props,
          container = this.container;
        var coord = props.coord;
        var records = this.mapping();
        var clip = this.getClip();
        return (0, _jsx.jsx)(View, (0, _objectSpread2.default)((0, _objectSpread2.default)({}, props), {}, {
          coord: coord,
          records: records,
          clip: clip
        }));
      }
    }]);
    return Point;
  }(_geometry.default);
};
exports.default = _default;
