// @antv/f2-graphic/es/engine/canvas.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import EventEmit from '../event/emit';
import EventController from '../event/controller';
import CanvasElement from './canvas-element';
import { mix, isString, substitute } from '@antv/util';
import * as DOMUtil from '../util/dom';
import Container from './container';
import Group from './group';
import { requestAnimationFrame } from '../util/requestAnimationFrame';
import lang from './lang';
var getPixelRatio = DOMUtil.getPixelRatio,
  getDomById = DOMUtil.getDomById,
  getWidth = DOMUtil.getWidth,
  getHeight = DOMUtil.getHeight,
  isCanvasElement = DOMUtil.isCanvasElement;
var Canvas = /*#__PURE__*/function (_EventEmit) {
  _inherits(Canvas, _EventEmit);
  var _super = _createSuper(Canvas);
  function Canvas(cfg) {
    var _this;
    _classCallCheck(this, Canvas);
    _this = _super.call(this);
    var title = cfg.title;
    var ariaLabel = title ? substitute(lang.general.withTitle, {
      title: title
    }) : lang.general.title;
    _this._attrs = mix({
      type: 'canvas',
      children: [],
      ariaLabel: ariaLabel
    }, cfg);
    _this._initPixelRatio();
    _this._initCanvas();
    return _this;
  }
  _createClass(Canvas, [{
    key: "get",
    value: /* eslint-enable */
    function get(name) {
      return this._attrs[name];
    }
  }, {
    key: "set",
    value: function set(name, value) {
      this._attrs[name] = value;
    }
  }, {
    key: "_initPixelRatio",
    value: function _initPixelRatio() {
      var pixelRatio = this.get('pixelRatio');
      if (!pixelRatio) {
        this.set('pixelRatio', getPixelRatio());
      }
    }
  }, {
    key: "beforeDraw",
    value: function beforeDraw() {
      var context = this._attrs.context;
      var el = this._attrs.el;
      context && context.clearRect && context.clearRect(0, 0, el.width, el.height);
    }
  }, {
    key: "_initCanvas",
    value: function _initCanvas() {
      var el = this.get('el');
      var context = this.get('context');
      if (!el && !context) {
        throw new Error('Please specify the id, el or context of the chart!');
      }
      var canvas;
      if (el) {
        // DOMElement or String
        canvas = isString(el) ? getDomById(el) : el;
      } else {
        // 说明没有指定el
        canvas = CanvasElement.create(context);
      }
      if (context && canvas && !canvas.getContext) {
        canvas.getContext = function () {
          return context;
        };
      }
      var width = this.get('width') || getWidth(canvas) || canvas.width;
      var height = this.get('height') || getHeight(canvas) || canvas.height;
      this.set('canvas', this);
      this.set('el', canvas);
      this.set('context', context || canvas.getContext('2d'));
      this.changeSize(width, height);
      // 初始化事件控制器
      var eventController = new EventController({
        canvas: this,
        el: canvas
      });
      this.set('eventController', eventController);
    }
  }, {
    key: "changeSize",
    value: function changeSize(width, height) {
      var pixelRatio = this.get('pixelRatio');
      var canvasDOM = this.get('el'); // HTMLCanvasElement or canvasElement
      // 浏览器环境设置style样式
      if (canvasDOM.style) {
        canvasDOM.style.width = width + 'px';
        canvasDOM.style.height = height + 'px';
      }
      if (isCanvasElement(canvasDOM)) {
        canvasDOM.width = width * pixelRatio;
        canvasDOM.height = height * pixelRatio;
        if (pixelRatio !== 1) {
          var ctx = this.get('context');
          ctx.scale(pixelRatio, pixelRatio);
        }
      }
      this.set('width', width);
      this.set('height', height);
    }
  }, {
    key: "getWidth",
    value: function getWidth() {
      var pixelRatio = this.get('pixelRatio');
      var width = this.get('width');
      return width * pixelRatio;
    }
  }, {
    key: "getHeight",
    value: function getHeight() {
      var pixelRatio = this.get('pixelRatio');
      var height = this.get('height');
      return height * pixelRatio;
    }
  }, {
    key: "getPointByClient",
    value: function getPointByClient(clientX, clientY) {
      var el = this.get('el');
      var bbox = el.getBoundingClientRect();
      var width = bbox.right - bbox.left;
      var height = bbox.bottom - bbox.top;
      return {
        x: (clientX - bbox.left) * (el.width / width),
        y: (clientY - bbox.top) * (el.height / height)
      };
    }
  }, {
    key: "_beginDraw",
    value: function _beginDraw() {
      this._attrs.toDraw = true;
    }
  }, {
    key: "_endDraw",
    value: function _endDraw() {
      this._attrs.toDraw = false;
      this.emit('afterdraw', {});
    }
  }, {
    key: "draw",
    value: function draw() {
      var _this2 = this;
      var drawInner = function drawInner() {
        _this2.set('animateHandler', requestAnimationFrame(function () {
          _this2.set('animateHandler', undefined);
          if (_this2.get('toDraw')) {
            drawInner();
          }
        }));
        _this2.beforeDraw();
        try {
          var context = _this2._attrs.context;
          _this2.drawChildren(context);
          // 支付宝，微信小程序，需要调context.draw才能完成绘制， 所以这里直接判断是否有.draw方法
          if (context.draw) {
            context.draw();
          }
          // 设置无障碍文本
          _this2.setAriaLabel();
        } catch (ev) {
          console.warn('error in draw canvas, detail as:');
          console.warn(ev);
          _this2._endDraw();
        }
        _this2._endDraw();
      };
      if (this.get('destroyed')) {
        return;
      }
      if (this.get('animateHandler')) {
        this._beginDraw();
      } else {
        drawInner();
      }
    }
    // 设置无障碍文本
  }, {
    key: "setAriaLabel",
    value: function setAriaLabel() {
      var el = this._attrs.el;
      var ariaLabel = this._getAriaLabel();
      if (ariaLabel && el.setAttribute) {
        el.setAttribute('aria-label', ariaLabel);
      }
    }
  }, {
    key: "destroy",
    value: function destroy() {
      if (this.get('destroyed')) {
        return;
      }
      // 需要清理 canvas 画布内容，否则会导致 spa 应用 ios 下 canvas 白屏
      // https://stackoverflow.com/questions/52532614/total-canvas-memory-use-exceeds-the-maximum-limit-safari-12
      // https://github.com/antvis/F2/issues/630
      var el = this.get('el');
      el.width = 0;
      el.height = 0;
      this.clear();
      this._attrs = {};
      this.set('destroyed', true);
    }
  }, {
    key: "isDestroyed",
    value: function isDestroyed() {
      return this.get('destroyed');
    }
  }]);
  return Canvas;
}(EventEmit); // @ts-ignore
mix(Canvas.prototype, Container, {
  getGroupClass: function getGroupClass() {
    return Group;
  }
});
export default Canvas;
