// @antv/f2-graphic/es/engine/group.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _get from "@babel/runtime/helpers/esm/get";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import { mix } from '@antv/util';
import Rect from './shape/rect';
import Container from './container';
import Vector2 from '../util/vector2';
var Group = /*#__PURE__*/function (_Rect) {
  _inherits(Group, _Rect);
  var _super = _createSuper(Group);
  function Group() {
    _classCallCheck(this, Group);
    return _super.apply(this, arguments);
  }
  _createClass(Group, [{
    key: "_initProperties",
    value: /* eslint-enable */
    function _initProperties() {
      this._attrs = {
        type: 'group',
        zIndex: 0,
        visible: true,
        destroyed: false,
        isGroup: true,
        canFill: true,
        canStroke: true,
        children: [],
        attrs: {
          x: 0,
          y: 0,
          width: 0,
          height: 0,
          radius: 0,
          lineWidth: 0
        }
      };
    }
  }, {
    key: "getBBox",
    value: function getBBox() {
      var minX = Infinity;
      var maxX = -Infinity;
      var minY = Infinity;
      var maxY = -Infinity;
      var children = this.get('children');
      for (var i = 0, length = children.length; i < length; i++) {
        var child = children[i];
        if (child.get('visible')) {
          var box = child.getBBox();
          if (!box) {
            continue;
          }
          var leftTop = [box.minX, box.minY];
          var leftBottom = [box.minX, box.maxY];
          var rightTop = [box.maxX, box.minY];
          var rightBottom = [box.maxX, box.maxY];
          var matrix = child.attr('matrix');
          Vector2.transformMat2d(leftTop, leftTop, matrix);
          Vector2.transformMat2d(leftBottom, leftBottom, matrix);
          Vector2.transformMat2d(rightTop, rightTop, matrix);
          Vector2.transformMat2d(rightBottom, rightBottom, matrix);
          minX = Math.min(leftTop[0], leftBottom[0], rightTop[0], rightBottom[0], minX);
          maxX = Math.max(leftTop[0], leftBottom[0], rightTop[0], rightBottom[0], maxX);
          minY = Math.min(leftTop[1], leftBottom[1], rightTop[1], rightBottom[1], minY);
          maxY = Math.max(leftTop[1], leftBottom[1], rightTop[1], rightBottom[1], maxY);
        }
      }
      return {
        minX: minX,
        minY: minY,
        maxX: maxX,
        maxY: maxY,
        x: minX,
        y: minY,
        width: maxX - minX,
        height: maxY - minY
      };
    }
  }, {
    key: "createPath",
    value: function createPath(context) {
      var attrs = this.get('attrs');
      // 只有在有fillStyle或strokeStyle 时才需要绘制
      if (!attrs.fillStyle && !attrs.strokeStyle) {
        return;
      }
      _get(_getPrototypeOf(Group.prototype), "createPath", this).call(this, context);
    }
  }, {
    key: "drawInner",
    value: function drawInner(context) {
      _get(_getPrototypeOf(Group.prototype), "drawInner", this).call(this, context);
      this.drawChildren(context);
    }
  }, {
    key: "destroy",
    value: function destroy() {
      if (this.get('destroyed')) {
        return;
      }
      this.clear();
      _get(_getPrototypeOf(Group.prototype), "destroy", this).call(this);
    }
  }]);
  return Group;
}(Rect); // @ts-ignore
mix(Group.prototype, Container, {
  getGroupClass: function getGroupClass() {
    return Group;
  }
});
export default Group;
