// @antv/f2/lib/components/guide/views/Line.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _jsx = require("../../../jsx");
var _util = require("@antv/util");
var _default = function _default(props, context) {
  var _props$theme = props.theme,
    theme = _props$theme === void 0 ? {} : _props$theme;
  var _deepMix = (0, _util.deepMix)((0, _objectSpread2.default)({}, theme.line), props),
    points = _deepMix.points,
    style = _deepMix.style,
    offsetX = _deepMix.offsetX,
    offsetY = _deepMix.offsetY,
    animation = _deepMix.animation;
  var _ref = points[0] || {},
    x1 = _ref.x,
    y1 = _ref.y;
  var _ref2 = points[1] || {},
    x2 = _ref2.x,
    y2 = _ref2.y;
  var offsetXNum = context.px2hd(offsetX);
  var offsetYNum = context.px2hd(offsetY);
  var posX1 = x1 + ((0, _util.isArray)(offsetXNum) ? offsetXNum[0] || 0 : offsetXNum || 0);
  var posY1 = y1 + ((0, _util.isArray)(offsetYNum) ? offsetYNum[0] || 0 : offsetYNum || 0);
  var posX2 = x2 + ((0, _util.isArray)(offsetXNum) ? offsetXNum[1] || 0 : offsetXNum || 0);
  var posY2 = y2 + ((0, _util.isArray)(offsetYNum) ? offsetYNum[1] || 0 : offsetYNum || 0);
  return (0, _jsx.jsx)("group", null, (0, _jsx.jsx)("line", {
    attrs: (0, _objectSpread2.default)({
      x1: posX1,
      y1: posY1,
      x2: posX2,
      y2: posY2
    }, style),
    animation: animation
  }));
};
exports.default = _default;
