// @antv/f2/lib/components/guide/views/Rect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _jsx = require("../../../jsx");
var _util = require("@antv/util");
var _default = function _default(props) {
  var _props$theme = props.theme,
    theme = _props$theme === void 0 ? {} : _props$theme;
  var _deepMix = (0, _util.deepMix)((0, _objectSpread2.default)({}, theme.rect), props),
    points = _deepMix.points,
    style = _deepMix.style,
    animation = _deepMix.animation;
  var start = points[0] || {};
  var end = points[1] || {};
  return (0, _jsx.jsx)("group", null, (0, _jsx.jsx)("rect", {
    attrs: (0, _objectSpread2.default)({
      x: Math.min(start.x, end.x),
      y: Math.min(start.y, end.y),
      width: Math.abs(end.x - start.x),
      height: Math.abs(start.y - end.y)
    }, style),
    animation: animation
  }));
};
exports.default = _default;
