// @antv/f2/lib/components/guide/views/Text.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _jsx = require("../../../jsx");
var _util = require("@antv/util");
var _default = function _default(props, context) {
  var _props$theme = props.theme,
    theme = _props$theme === void 0 ? {} : _props$theme;
  var _deepMix = (0, _util.deepMix)((0, _objectSpread2.default)({}, theme.text), props),
    points = _deepMix.points,
    style = _deepMix.style,
    offsetX = _deepMix.offsetX,
    offsetY = _deepMix.offsetY,
    content = _deepMix.content,
    animation = _deepMix.animation;
  var _ref = points[0] || {},
    x = _ref.x,
    y = _ref.y;
  var offsetXNum = context.px2hd(offsetX);
  var offsetYNum = context.px2hd(offsetY);
  var posX = x + (offsetXNum || 0);
  var posY = y + (offsetYNum || 0);
  return (0, _jsx.jsx)("text", {
    attrs: (0, _objectSpread2.default)({
      text: content,
      x: posX,
      y: posY
    }, style),
    animation: (0, _util.deepMix)({
      update: {
        easing: 'linear',
        duration: 450,
        property: ['x', 'y']
      }
    }, animation)
  });
};
exports.default = _default;
