// @antv/f2/lib/attr/base.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _util = require("@antv/util");
var Base = /*#__PURE__*/function () {
  function Base(options) {
    (0, _classCallCheck2.default)(this, Base);
    (0, _util.mix)(this, options);
    var scale = this.scale,
      field = this.field,
      data = this.data;
    if (!scale && data) {
      var values = (0, _util.valuesOfKey)(data, field);
      this.scale = this.createScale({
        values: values,
        field: field
      });
    }
  }
  (0, _createClass2.default)(Base, [{
    key: "createScale",
    value: function createScale(_scaleConfig) {
      return null;
    }
    // 数据映射方法
  }, {
    key: "_mapping",
    value: function _mapping(value) {
      return value;
    }
  }, {
    key: "update",
    value: function update(options) {
      (0, _util.mix)(this, options);
    }
  }, {
    key: "setRange",
    value: function setRange(range) {
      this.range = range;
    }
    // 归一化，参数是原始数据，返回是归一化的数据
  }, {
    key: "normalize",
    value: function normalize(value) {
      var scale = this.scale;
      if ((0, _util.isArray)(value)) {
        return value.map(function (v) {
          return scale.scale(v);
        });
      }
      return scale.scale(value);
    }
    // convert 参数是归一化的数据，返回定义域的值
  }, {
    key: "convert",
    value: function convert(value) {
      return value;
    }
    // 等于 normalize + convert， 参数是原始数据，返回是定义域的值
  }, {
    key: "mapping",
    value: function mapping(value) {
      var child = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
      var rst = (0, _util.isFunction)(this.callback) ? this.callback(value, child) : null;
      if (!(0, _util.isNil)(rst)) {
        return rst;
      }
      return this._mapping(value);
    }
  }]);
  return Base;
}();
var _default = Base;
exports.default = _default;
