// @antv/f2/lib/components/guide/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TextGuide = exports.TagGuide = exports.RectGuide = exports.PointGuide = exports.LineGuide = exports.ImageGuide = exports.ArcGuide = void 0;
Object.defineProperty(exports, "withGuide", {
  enumerable: true,
  get: function get() {
    return _withGuide.default;
  }
});
var _withGuide = _interopRequireDefault(require("./withGuide"));
var _Text = _interopRequireDefault(require("./views/Text"));
var _Point = _interopRequireDefault(require("./views/Point"));
var _Line = _interopRequireDefault(require("./views/Line"));
var _Arc = _interopRequireDefault(require("./views/Arc"));
var _Rect = _interopRequireDefault(require("./views/Rect"));
var _Image = _interopRequireDefault(require("./views/Image"));
var _Tag = _interopRequireDefault(require("./views/Tag"));
var DefaultGuideView = function DefaultGuideView() {
  return null;
};
var TextGuide = (0, _withGuide.default)(_Text.default);
exports.TextGuide = TextGuide;
var PointGuide = (0, _withGuide.default)(_Point.default);
exports.PointGuide = PointGuide;
var LineGuide = (0, _withGuide.default)(_Line.default);
exports.LineGuide = LineGuide;
var ArcGuide = (0, _withGuide.default)(_Arc.default);
exports.ArcGuide = ArcGuide;
var RectGuide = (0, _withGuide.default)(_Rect.default);
exports.RectGuide = RectGuide;
var ImageGuide = (0, _withGuide.default)(_Image.default);
exports.ImageGuide = ImageGuide;
var TagGuide = (0, _withGuide.default)(_Tag.default);
exports.TagGuide = TagGuide;
var _default = (0, _withGuide.default)(DefaultGuideView);
exports.default = _default;
