// @antv/f2/lib/components/sunburst/sunburstView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _jsx = require("../../jsx");
var _default = function _default(props) {
  var coord = props.coord,
    node = props.node;
  var children = node.children;
  var _coord$center = coord.center,
    x = _coord$center.x,
    y = _coord$center.y;
  var renderNodes = function renderNodes(nodes) {
    return (0, _jsx.jsx)("group", null, nodes.map(function (node) {
      var xMin = node.xMin,
        xMax = node.xMax,
        yMin = node.yMin,
        yMax = node.yMax,
        color = node.color,
        children = node.children;
      return (0, _jsx.jsx)("group", null, (0, _jsx.jsx)("sector", {
        attrs: {
          x: x,
          y: y,
          lineWidth: '1px',
          stroke: '#fff',
          startAngle: xMin,
          endAngle: xMax,
          r0: yMin,
          r: yMax,
          anticlockwise: false,
          fill: color
        }
      }), children && children.length ? renderNodes(children) : null);
    }));
  };
  return renderNodes(children);
};
exports.default = _default;
