// @nutui/nutui/dist/packages/_es/Address.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { ref, computed, reactive, watch, toRefs, h, nextTick, resolveComponent, openBlock, createBlock, withCtx, createElementVNode, withDirectives, createVNode, mergeProps, vShow, toDisplayString, createCommentVNode, createElementBlock, Fragment, renderList, normalizeClass, normalizeStyle, resolveDynamicComponent, normalizeProps, createTextVNode, guardReactiveProps, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { p as popupProps } from "./props.js";
import _sfc_main$1 from "./Icon.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
import "./pxCheck.js";
const { componentName, create, translate } = createComponent("address");
const _sfc_main = create({
  inheritAttrs: false,
  props: {
    ...popupProps,
    modelValue: {
      type: Array,
      default: () => []
    },
    type: {
      type: String,
      default: "custom"
    },
    customAddressTitle: {
      type: String,
      default: ""
    },
    province: {
      type: Array,
      default: () => []
    },
    city: {
      type: Array,
      default: () => []
    },
    country: {
      type: Array,
      default: () => []
    },
    town: {
      type: Array,
      default: () => []
    },
    isShowCustomAddress: {
      type: Boolean,
      default: true
    },
    existAddress: {
      type: Array,
      default: () => []
    },
    existAddressTitle: {
      type: String,
      default: ""
    },
    customAndExistTitle: {
      type: String,
      default: ""
    },
    defaultIcon: {
      type: String,
      default: "location2"
    },
    selectedIcon: {
      type: String,
      default: "Check"
    },
    closeBtnIcon: {
      type: String,
      default: "circle-close"
    },
    backBtnIcon: {
      type: String,
      default: "left"
    },
    height: {
      type: [String, Number],
      default: "200px"
    },
    columnsPlaceholder: {
      type: [String, Array],
      default: ""
    }
  },
  emits: ["update:visible", "update:modelValue", "type", "change", "selected", "close", "close-mask", "switch-module"],
  setup(props, { emit }) {
    const regionLine = ref(null);
    const tabRegion = ref(null);
    const showPopup = ref(props.visible);
    const privateType = ref(props.type);
    const tabIndex = ref(0);
    const tabName = ref(["province", "city", "country", "town"]);
    const regionList = computed(() => {
      switch (tabIndex.value) {
        case 0:
          return props.province;
        case 1:
          return props.city;
        case 2:
          return props.country;
        default:
          return props.town;
      }
    });
    const renderIcon = (n) => {
      return h(_sfc_main$1, {
        class: `${componentName}-select-icon`,
        type: "self",
        size: "13px",
        name: n
      });
    };
    const transformData = (data) => {
      if (!Array.isArray(data))
        throw new TypeError("params muse be array.");
      if (!data.length)
        return [];
      data.forEach((item) => {
        if (!item.title) {
          console.error("[NutUI] <Address> \u8BF7\u68C0\u67E5\u6570\u7EC4\u9009\u9879\u7684 title \u503C\u662F\u5426\u6709\u8BBE\u7F6E ,title \u4E3A\u5FC5\u586B\u9879 .");
          return;
        }
      });
      const newData = [];
      data = data.sort((a, b) => a.title.localeCompare(b.title));
      data.forEach((item) => {
        const index2 = newData.findIndex((value) => value.title === item.title);
        if (index2 <= -1) {
          newData.push({
            title: item.title,
            list: [].concat(item)
          });
        } else {
          newData[index2].list.push(item);
        }
      });
      return newData;
    };
    let selectedRegion = ref([]);
    let selectedExistAddress = reactive({});
    const closeWay = ref("self");
    const lineDistance = ref(20);
    const initCustomSelected = () => {
      const defaultValue = props.modelValue;
      const num = defaultValue.length;
      if (num > 0) {
        tabIndex.value = num - 1;
        if (regionList.value.length == 0) {
          tabIndex.value = 0;
          return;
        }
        for (let index2 = 0; index2 < num; index2++) {
          let arr = [];
          switch (index2) {
            case 0:
              arr = props.province;
              break;
            case 1:
              arr = props.city;
              break;
            case 2:
              arr = props.country;
              break;
            default:
              arr = props.town;
          }
          selectedRegion.value[index2] = arr.filter((item) => item.id == defaultValue[index2])[0];
        }
        lineAnimation();
      }
    };
    const getTabName = (item, index2) => {
      if (item && item.name)
        return item.name;
      if (tabIndex.value < index2 && item) {
        return item.name;
      } else {
        return props.columnsPlaceholder[index2] || translate("select");
      }
    };
    const lineAnimation = () => {
      nextTick(() => {
        const name = tabRegion.value && tabRegion.value.getElementsByClassName("active")[0];
        if (name) {
          const distance = name.offsetLeft;
          lineDistance.value = distance ? distance : 20;
        }
      });
    };
    const nextAreaList = (item) => {
      const tab = tabIndex.value;
      const callBackParams = {
        custom: tabName.value[tab]
      };
      selectedRegion.value[tab] = item;
      for (let i = tab + 2; i < 4; i++) {
        selectedRegion.value.splice(i, 1);
      }
      if (tab < 3) {
        tabIndex.value = tab + 1;
        lineAnimation();
        callBackParams.next = tabName.value[tabIndex.value];
        callBackParams.value = item;
        emit("change", callBackParams);
      } else {
        handClose();
        emit("update:modelValue");
      }
    };
    const changeRegionTab = (item, index2) => {
      if (getTabName(item, index2)) {
        tabIndex.value = index2;
        lineAnimation();
      }
    };
    const selectedExist = (item) => {
      const copyExistAdd = props.existAddress;
      let prevExistAdd = {};
      copyExistAdd.forEach((list) => {
        if (list && list.selectedAddress)
          prevExistAdd = list;
        list.selectedAddress = false;
      });
      item.selectedAddress = true;
      selectedExistAddress = item;
      emit("selected", prevExistAdd, item, copyExistAdd);
      handClose();
    };
    const initAddress = () => {
      selectedRegion.value = [];
      tabIndex.value = 0;
      lineAnimation();
    };
    const handClose = (type = "self") => {
      if (!props.closeBtnIcon)
        return;
      closeWay.value = type == "cross" ? "cross" : "self";
      showPopup.value = false;
    };
    const clickOverlay = () => {
      closeWay.value = "mask";
    };
    const close = () => {
      const data = {
        addressIdStr: "",
        addressStr: "",
        province: selectedRegion.value[0],
        city: selectedRegion.value[1],
        country: selectedRegion.value[2],
        town: selectedRegion.value[3]
      };
      const callBackParams = {
        data: {},
        type: privateType.value
      };
      if (["custom", "custom2"].includes(privateType.value)) {
        [0, 1, 2, 3].forEach((i) => {
          const item = selectedRegion.value[i];
          data.addressIdStr += `${i ? "_" : ""}${item && item.id || 0}`;
          data.addressStr += item && item.name || "";
        });
        callBackParams.data = data;
      } else {
        callBackParams.data = selectedExistAddress;
      }
      initAddress();
      if (closeWay.value == "self") {
        emit("close", callBackParams);
      } else {
        emit("close-mask", { closeWay });
      }
      emit("update:visible", false);
    };
    const switchModule = () => {
      const type = privateType.value;
      privateType.value = type == "exist" ? "custom" : "exist";
      initAddress();
      emit("switch-module", { type: privateType.value });
    };
    const handleElevatorItem = (key, item) => {
      nextAreaList(item);
    };
    watch(
      () => props.visible,
      (value) => {
        showPopup.value = value;
      }
    );
    watch(
      () => showPopup.value,
      (value) => {
        if (value) {
          initCustomSelected();
        }
      }
    );
    return {
      showPopup,
      privateType,
      tabIndex,
      tabName,
      selectedRegion,
      switchModule,
      closeWay,
      close,
      getTabName,
      nextAreaList,
      regionLine,
      tabRegion,
      lineDistance,
      changeRegionTab,
      selectedExist,
      clickOverlay,
      handClose,
      handleElevatorItem,
      initCustomSelected,
      ...toRefs(props),
      translate,
      regionList,
      transformData,
      renderIcon
    };
  }
});
const _hoisted_1 = { class: "nut-address" };
const _hoisted_2 = { class: "nut-address__header" };
const _hoisted_3 = { class: "nut-address__header__title" };
const _hoisted_4 = {
  key: 0,
  class: "custom-address"
};
const _hoisted_5 = {
  class: "nut-address-region-tab",
  ref: "tabRegion"
};
const _hoisted_6 = ["onClick"];
const _hoisted_7 = {
  key: 0,
  class: "active tab-item"
};
const _hoisted_8 = {
  key: 0,
  class: "region-con"
};
const _hoisted_9 = { class: "region-group" };
const _hoisted_10 = ["onClick"];
const _hoisted_11 = {
  key: 1,
  class: "elevator-group"
};
const _hoisted_12 = {
  key: 1,
  class: "exist-address"
};
const _hoisted_13 = { class: "exist-address-group" };
const _hoisted_14 = { class: "exist-ul" };
const _hoisted_15 = ["onClick"];
const _hoisted_16 = { class: "exist-item-info" };
const _hoisted_17 = {
  key: 0,
  class: "exist-item-info-name"
};
const _hoisted_18 = {
  key: 1,
  class: "exist-item-info-phone"
};
const _hoisted_19 = { class: "exist-item-info-bottom" };
const _hoisted_20 = { class: "btn" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  const _component_nut_elevator = resolveComponent("nut-elevator");
  const _component_nut_popup = resolveComponent("nut-popup");
  return openBlock(), createBlock(_component_nut_popup, {
    position: "bottom",
    "lock-scroll": _ctx.lockScroll,
    onClose: _ctx.close,
    onClickOverlay: _ctx.clickOverlay,
    onOpen: _cache[3] || (_cache[3] = ($event) => _ctx.closeWay = "self"),
    visible: _ctx.showPopup,
    "onUpdate:visible": _cache[4] || (_cache[4] = ($event) => _ctx.showPopup = $event),
    teleportDisable: _ctx.teleportDisable,
    teleport: _ctx.teleport
  }, {
    default: withCtx(() => [
      createElementVNode("view", _hoisted_1, [
        createElementVNode("view", _hoisted_2, [
          createElementVNode("view", {
            class: "arrow-back",
            onClick: _cache[0] || (_cache[0] = (...args) => _ctx.switchModule && _ctx.switchModule(...args))
          }, [
            withDirectives(createVNode(_component_nut_icon, mergeProps(_ctx.$attrs, {
              name: _ctx.backBtnIcon,
              color: "#cccccc"
            }), null, 16, ["name"]), [
              [vShow, _ctx.type == "exist" && _ctx.privateType == "custom" && _ctx.backBtnIcon]
            ])
          ]),
          createElementVNode("view", _hoisted_3, toDisplayString(_ctx.privateType == "custom" ? _ctx.customAddressTitle || _ctx.translate("selectRegion") : _ctx.existAddressTitle || _ctx.translate("deliveryTo")), 1),
          createElementVNode("view", {
            class: "arrow-close",
            onClick: _cache[1] || (_cache[1] = ($event) => _ctx.handClose("cross"))
          }, [
            _ctx.closeBtnIcon ? (openBlock(), createBlock(_component_nut_icon, mergeProps({ key: 0 }, _ctx.$attrs, {
              name: _ctx.closeBtnIcon,
              color: "#cccccc",
              size: "18px"
            }), null, 16, ["name"])) : createCommentVNode("", true)
          ])
        ]),
        ["custom", "custom2"].includes(_ctx.privateType) ? (openBlock(), createElementBlock("view", _hoisted_4, [
          createElementVNode("view", _hoisted_5, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.selectedRegion, (item, index2) => {
              return openBlock(), createElementBlock("view", {
                class: normalizeClass(["tab-item", index2 == _ctx.tabIndex ? "active" : ""]),
                key: index2,
                onClick: ($event) => _ctx.changeRegionTab(item, index2)
              }, [
                createElementVNode("view", null, toDisplayString(_ctx.getTabName(item, index2)), 1)
              ], 10, _hoisted_6);
            }), 128)),
            _ctx.tabIndex == _ctx.selectedRegion.length ? (openBlock(), createElementBlock("view", _hoisted_7, [
              createElementVNode("view", null, toDisplayString(_ctx.getTabName(null, _ctx.selectedRegion.length)), 1)
            ])) : createCommentVNode("", true),
            createElementVNode("view", {
              class: "region-tab-line",
              ref: "regionLine",
              style: normalizeStyle({ left: _ctx.lineDistance + "px" })
            }, null, 4)
          ], 512),
          _ctx.privateType == "custom" ? (openBlock(), createElementBlock("view", _hoisted_8, [
            createElementVNode("ul", _hoisted_9, [
              (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.regionList, (item, index2) => {
                var _a, _b;
                return openBlock(), createElementBlock("li", {
                  key: index2,
                  class: normalizeClass(["region-item", ((_a = _ctx.selectedRegion[_ctx.tabIndex]) == null ? void 0 : _a.id) == item.id ? "active" : ""]),
                  onClick: ($event) => _ctx.nextAreaList(item)
                }, [
                  createElementVNode("div", null, [
                    ((_b = _ctx.selectedRegion[_ctx.tabIndex]) == null ? void 0 : _b.id) == item.id ? (openBlock(), createBlock(resolveDynamicComponent(_ctx.renderIcon(_ctx.selectedIcon)), normalizeProps(mergeProps({ key: 0 }, _ctx.$attrs)), null, 16)) : createCommentVNode("", true),
                    createTextVNode(" " + toDisplayString(item.name), 1)
                  ])
                ], 10, _hoisted_10);
              }), 128))
            ])
          ])) : (openBlock(), createElementBlock("view", _hoisted_11, [
            createVNode(_component_nut_elevator, {
              height: _ctx.height,
              "index-list": _ctx.transformData(_ctx.regionList),
              onClickItem: _ctx.handleElevatorItem
            }, null, 8, ["height", "index-list", "onClickItem"])
          ]))
        ])) : (openBlock(), createElementBlock("view", _hoisted_12, [
          createElementVNode("div", _hoisted_13, [
            createElementVNode("ul", _hoisted_14, [
              (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.existAddress, (item, index2) => {
                return openBlock(), createElementBlock("li", {
                  class: normalizeClass(["exist-item", item.selectedAddress ? "active" : ""]),
                  key: index2,
                  onClick: ($event) => _ctx.selectedExist(item)
                }, [
                  (openBlock(), createBlock(resolveDynamicComponent(_ctx.renderIcon(item.selectedAddress ? _ctx.selectedIcon : _ctx.defaultIcon)), normalizeProps(guardReactiveProps(_ctx.$attrs)), null, 16)),
                  createElementVNode("div", _hoisted_16, [
                    item.name ? (openBlock(), createElementBlock("div", _hoisted_17, toDisplayString(item.name), 1)) : createCommentVNode("", true),
                    item.phone ? (openBlock(), createElementBlock("div", _hoisted_18, toDisplayString(item.phone), 1)) : createCommentVNode("", true),
                    createElementVNode("div", _hoisted_19, [
                      createElementVNode("view", null, toDisplayString(item.provinceName + item.cityName + item.countyName + item.townName + item.addressDetail), 1)
                    ])
                  ])
                ], 10, _hoisted_15);
              }), 128))
            ])
          ]),
          _ctx.isShowCustomAddress ? (openBlock(), createElementBlock("div", {
            key: 0,
            class: "choose-other",
            onClick: _cache[2] || (_cache[2] = (...args) => _ctx.switchModule && _ctx.switchModule(...args))
          }, [
            createElementVNode("div", _hoisted_20, toDisplayString(_ctx.customAndExistTitle || _ctx.translate("chooseAnotherAddress")), 1)
          ])) : createCommentVNode("", true)
        ])),
        renderSlot(_ctx.$slots, "bottom")
      ])
    ]),
    _: 3
  }, 8, ["lock-scroll", "onClose", "onClickOverlay", "visible", "teleportDisable", "teleport"]);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
