// @nutui/nutui/dist/packages/_es/AvatarGroup.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { toRefs, ref, computed, onMounted, provide, openBlock, createElementBlock, normalizeClass, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("avatar-group");
const _sfc_main = create({
  props: {
    maxContent: {
      type: String,
      default: ""
    },
    maxCount: {
      type: [Number, String],
      default: ""
    },
    maxBgColor: {
      type: String,
      default: "#eee"
    },
    maxColor: {
      type: String,
      default: "#666"
    },
    size: {
      type: String,
      default: ""
    },
    shape: {
      type: String,
      default: ""
    },
    span: {
      type: String,
      default: "-8"
    },
    zIndex: {
      type: String,
      default: "left"
    }
  },
  setup(props, { emit, slots }) {
    toRefs(props);
    const avatarGroupRef = ref(null);
    const index2 = ref(0);
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    onMounted(() => {
    });
    provide("avatarGroup", {
      props,
      avatarGroupRef,
      index: index2
    });
    return {
      classes,
      avatarGroupRef
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    ref: "avatarGroupRef"
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 2);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
