// @nutui/nutui/dist/packages/_es/CategoryPane.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { c as createComponent } from "./component.js";
import { openBlock, createElementBlock, Fragment, renderList, createElementVNode, toDisplayString, createCommentVNode } from "vue";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("categorypane");
const _sfc_main = create({
  props: {
    type: {
      type: String,
      default: "classify"
    },
    categoryChild: {
      type: Array,
      default: () => []
    },
    customCategory: {
      type: Array,
      default: () => []
    }
  },
  setup(props, { emit }) {
    const onChange = (sku) => {
      emit("onChange", sku);
    };
    return {
      onChange
    };
  }
});
const _hoisted_1 = { class: "nut-categorypane" };
const _hoisted_2 = {
  key: 0,
  class: "nut-categorypane__cateListRight"
};
const _hoisted_3 = { class: "nut-categorypane__childTitle" };
const _hoisted_4 = {
  key: 0,
  class: "nut-categorypane__childItemList"
};
const _hoisted_5 = ["onClick"];
const _hoisted_6 = ["src"];
const _hoisted_7 = { class: "nut-categorypane__skuImg" };
const _hoisted_8 = {
  key: 1,
  class: "nut-categorypane__cateListRight"
};
const _hoisted_9 = { class: "nut-categorypane__childTitle" };
const _hoisted_10 = {
  key: 0,
  class: "nut-categorypane__childItemList"
};
const _hoisted_11 = ["onClick"];
const _hoisted_12 = { class: "nut-categorypane__skuName" };
const _hoisted_13 = {
  key: 2,
  class: "nut-categorypane__selfItemList"
};
const _hoisted_14 = ["onClick"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", _hoisted_1, [
    _ctx.type == "classify" ? (openBlock(), createElementBlock("div", _hoisted_2, [
      (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.categoryChild, (item, index2) => {
        return openBlock(), createElementBlock("div", { key: index2 }, [
          createElementVNode("div", _hoisted_3, toDisplayString(item == null ? void 0 : item.catName), 1),
          (item == null ? void 0 : item.catType) == 1 ? (openBlock(), createElementBlock("div", _hoisted_4, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(item.childCateList, (sku, key) => {
              return openBlock(), createElementBlock("div", {
                class: "nut-categorypane__childItem",
                key,
                onClick: ($event) => _ctx.onChange(sku)
              }, [
                createElementVNode("img", {
                  class: "nut-categorypane__childImg",
                  src: sku.backImg
                }, null, 8, _hoisted_6),
                createElementVNode("div", _hoisted_7, toDisplayString(sku == null ? void 0 : sku.catName), 1)
              ], 8, _hoisted_5);
            }), 128))
          ])) : createCommentVNode("", true)
        ]);
      }), 128))
    ])) : createCommentVNode("", true),
    _ctx.type == "text" ? (openBlock(), createElementBlock("div", _hoisted_8, [
      (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.categoryChild, (item, index2) => {
        return openBlock(), createElementBlock("div", { key: index2 }, [
          createElementVNode("div", _hoisted_9, toDisplayString(item == null ? void 0 : item.catName), 1),
          (item == null ? void 0 : item.catType) == 1 ? (openBlock(), createElementBlock("div", _hoisted_10, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(item.childCateList, (sku, key) => {
              return openBlock(), createElementBlock("div", {
                class: "nut-categorypane__childItem",
                key,
                onClick: ($event) => _ctx.onChange(sku)
              }, [
                createElementVNode("div", _hoisted_12, toDisplayString(sku == null ? void 0 : sku.catName), 1)
              ], 8, _hoisted_11);
            }), 128))
          ])) : createCommentVNode("", true)
        ]);
      }), 128))
    ])) : createCommentVNode("", true),
    _ctx.type == "custom" ? (openBlock(), createElementBlock("div", _hoisted_13, [
      (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.customCategory, (sku, key) => {
        return openBlock(), createElementBlock("div", {
          class: "nut-categorypane__skuName",
          key,
          onClick: ($event) => _ctx.onChange(sku)
        }, toDisplayString(sku == null ? void 0 : sku.catName), 9, _hoisted_14);
      }), 128))
    ])) : createCommentVNode("", true)
  ]);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
