// @nutui/nutui/dist/packages/_es/Category.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { ref, onMounted, openBlock, createElementBlock, createElementVNode, Fragment, renderList, normalizeClass, toDisplayString, createCommentVNode, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("category");
const _sfc_main = create({
  props: {
    type: {
      type: String,
      default: "classify"
    },
    category: {
      type: Array,
      default: () => []
    }
  },
  setup(props, { emit, slots }) {
    const checkIndex = ref(0);
    const categoryLeft = ref(false);
    const getChildList = (index2) => {
      checkIndex.value = index2;
      emit("change", index2);
    };
    onMounted(() => {
    });
    return {
      getChildList,
      checkIndex,
      categoryLeft
    };
  }
});
const _hoisted_1 = { class: "nut-category" };
const _hoisted_2 = { class: "nut-category__cateList" };
const _hoisted_3 = { key: 0 };
const _hoisted_4 = ["onClick"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", _hoisted_1, [
    createElementVNode("div", _hoisted_2, [
      _ctx.type == "classify" || _ctx.type == "text" ? (openBlock(), createElementBlock("div", _hoisted_3, [
        (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.category, (item, index2) => {
          return openBlock(), createElementBlock("div", {
            class: "nut-category__cateListLeft",
            key: index2
          }, [
            createElementVNode("div", {
              class: normalizeClass([_ctx.checkIndex == index2 ? "nut-category__cateListItemChecked" : "nut-category__cateListItem"]),
              onClick: ($event) => _ctx.getChildList(index2)
            }, toDisplayString(item == null ? void 0 : item.catName), 11, _hoisted_4)
          ]);
        }), 128))
      ])) : createCommentVNode("", true),
      renderSlot(_ctx.$slots, "default")
    ])
  ]);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
