// @nutui/nutui/dist/packages/_es/Cell.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { resolveComponent, openBlock, createElementBlock, normalizeClass, normalizeStyle, renderSlot, createBlock, mergeProps, createCommentVNode, Fragment, createElementVNode, toDisplayString, createTextVNode, computed } from "vue";
import { c as createComponent } from "./component.js";
import { u as useRouter } from "./index.js";
import { p as pxCheck } from "./pxCheck.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const _hoisted_1 = {
  key: 0,
  class: "nut-cell__icon"
};
const _hoisted_2 = {
  key: 1,
  class: "nut-cell__title"
};
const _hoisted_3 = { class: "title" };
const _hoisted_4 = { class: "nut-cell__title-desc" };
function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    style: normalizeStyle(_ctx.baseStyle),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    renderSlot(_ctx.$slots, "default", {}, () => [
      _ctx.icon || _ctx.$slots.icon ? (openBlock(), createElementBlock("view", _hoisted_1, [
        renderSlot(_ctx.$slots, "icon", {}, () => [
          _ctx.icon ? (openBlock(), createBlock(_component_nut_icon, mergeProps({ key: 0 }, _ctx.$attrs, {
            class: "icon",
            name: _ctx.icon
          }), null, 16, ["name"])) : createCommentVNode("", true)
        ])
      ])) : createCommentVNode("", true),
      _ctx.title || _ctx.subTitle || _ctx.$slots.title ? (openBlock(), createElementBlock("view", _hoisted_2, [
        _ctx.subTitle ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
          renderSlot(_ctx.$slots, "title", {}, () => [
            createElementVNode("view", _hoisted_3, toDisplayString(_ctx.title), 1)
          ]),
          createElementVNode("view", _hoisted_4, toDisplayString(_ctx.subTitle), 1)
        ], 64)) : renderSlot(_ctx.$slots, "title", { key: 1 }, () => [
          createTextVNode(toDisplayString(_ctx.title), 1)
        ])
      ])) : createCommentVNode("", true),
      _ctx.desc ? (openBlock(), createElementBlock("view", {
        key: 2,
        class: normalizeClass(["nut-cell__value", { "nut-cell__value--alone": !_ctx.title && !_ctx.subTitle && !_ctx.$slots.title }]),
        style: normalizeStyle({ "text-align": _ctx.descTextAlign })
      }, toDisplayString(_ctx.desc), 7)) : createCommentVNode("", true),
      renderSlot(_ctx.$slots, "link", {}, () => [
        _ctx.isLink || _ctx.to ? (openBlock(), createBlock(_component_nut_icon, mergeProps({
          key: 0,
          class: "nut-cell__link"
        }, _ctx.$attrs, { name: _ctx.rightIcon }), null, 16, ["name"])) : createCommentVNode("", true)
      ])
    ])
  ], 6);
}
const { componentName, create } = createComponent("cell");
const _sfc_main = create({
  props: {
    title: { type: String, default: "" },
    subTitle: { type: String, default: "" },
    desc: { type: String, default: "" },
    descTextAlign: { type: String, default: "right" },
    isLink: { type: Boolean, default: false },
    to: [String, Object],
    replace: { type: Boolean, default: false },
    roundRadius: { type: [String, Number], default: "" },
    url: { type: String, default: "" },
    icon: { type: String, default: "" },
    rightIcon: { type: String, default: "right" },
    center: { type: Boolean, default: false },
    size: { type: String, default: "" }
  },
  emits: ["click"],
  setup(props, { emit }) {
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true,
        [`${prefixCls}--clickable`]: props.isLink || props.to,
        [`${prefixCls}--center`]: props.center,
        [`${prefixCls}--large`]: props.size == "large"
      };
    });
    const router = useRouter();
    const baseStyle = computed(() => {
      return {
        borderRadius: pxCheck(props.roundRadius)
      };
    });
    const handleClick = (event) => {
      emit("click", event);
      if (props.to && router) {
        router[props.replace ? "replace" : "push"](props.to);
      } else if (props.url) {
        props.replace ? location.replace(props.url) : location.href = props.url;
      }
    };
    return {
      handleClick,
      classes,
      baseStyle
    };
  }
});
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", render]]);
export { index as default };
