// @nutui/nutui/dist/packages/_es/Divider.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { computed, openBlock, createElementBlock, normalizeClass, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("divider");
const _sfc_main = create({
  props: {
    contentPosition: {
      type: String,
      default: "center"
    },
    dashed: {
      type: Boolean,
      default: false
    },
    hairline: {
      type: Boolean,
      default: true
    },
    direction: {
      type: String,
      default: "horizontal"
    }
  },
  components: {},
  setup(props, context) {
    const classes = computed(() => {
      const prefixCls = componentName;
      let defaultClassesObj = {
        [prefixCls]: true
      };
      let classesObj = {};
      if (props.direction === "horizontal") {
        classesObj = {
          ...defaultClassesObj,
          [`${prefixCls}-center`]: context.slots.default,
          [`${prefixCls}-left`]: props.contentPosition === "left",
          [`${prefixCls}-right`]: props.contentPosition === "right",
          [`${prefixCls}-dashed`]: props.dashed,
          [`${prefixCls}-hairline`]: props.hairline
        };
      } else {
        classesObj = {
          ...defaultClassesObj,
          [`${prefixCls}-vertical`]: props.direction === "vertical"
        };
      }
      return classesObj;
    });
    return { classes };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return _ctx.direction === "horizontal" ? (openBlock(), createElementBlock("view", {
    key: 0,
    class: normalizeClass(_ctx.classes)
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 2)) : (openBlock(), createElementBlock("view", {
    key: 1,
    class: normalizeClass(_ctx.classes)
  }, null, 2));
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
