// @nutui/nutui/dist/packages/_es/Ellipsis.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { ref, reactive, computed, watch, onMounted, toRefs, openBlock, createElementBlock, normalizeClass, toDisplayString, createCommentVNode, createTextVNode, withModifiers } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("ellipsis");
const _sfc_main = create({
  props: {
    content: {
      type: String,
      default: ""
    },
    direction: {
      type: String,
      default: "end"
    },
    rows: {
      type: [Number, String],
      default: 1
    },
    expandText: {
      type: String,
      default: ""
    },
    collapseText: {
      type: String,
      default: ""
    },
    symbol: {
      type: String,
      default: "..."
    },
    lineHeight: {
      type: [Number, String],
      default: "20"
    }
  },
  emits: ["click", "change"],
  setup(props, { emit }) {
    const root = ref(null);
    let container = null;
    let maxHeight = 0;
    const ellipsis = ref();
    const state = reactive({
      exceeded: false,
      expanded: false
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    watch(
      () => props.content,
      (newV, oldVal) => {
        if (newV != oldVal) {
          createContainer();
        }
      }
    );
    onMounted(() => {
      createContainer();
    });
    const createContainer = () => {
      if (!root.value)
        return;
      const originStyle = window.getComputedStyle(root.value);
      container = document.createElement("div");
      const styleNames = Array.prototype.slice.apply(originStyle);
      styleNames.forEach((name) => {
        container.style.setProperty(name, originStyle.getPropertyValue(name));
      });
      container.style.position = "fixed";
      container.style.left = "999999px";
      container.style.top = "999999px";
      container.style.zIndex = "-1000";
      container.style.height = "auto";
      container.style.minHeight = "auto";
      container.style.maxHeight = "auto";
      container.style.textOverflow = "clip";
      container.style.whiteSpace = "normal";
      container.style.webkitLineClamp = "unset";
      container.style.display = "block";
      const lineHeight = pxToNumber(originStyle.lineHeight === "normal" ? props.lineHeight : originStyle.lineHeight);
      maxHeight = Math.floor(
        lineHeight * (Number(props.rows) + 0.5) + pxToNumber(originStyle.paddingTop) + pxToNumber(originStyle.paddingBottom)
      );
      container.innerText = props.content;
      document.body.appendChild(container);
      calcEllipse();
    };
    const calcEllipse = () => {
      if (container.offsetHeight <= maxHeight) {
        state.exceeded = false;
        document.body.removeChild(container);
      } else {
        state.exceeded = true;
        const end = props.content.length;
        const middle = Math.floor((0 + end) / 2);
        const ellipsised = props.direction === "middle" ? tailorMiddle([0, middle], [middle, end]) : tailor(0, end);
        ellipsis.value = ellipsised;
        document.body.removeChild(container);
      }
    };
    const tailor = (left, right) => {
      const actionText = state.expanded ? props.collapseText : props.expandText;
      const end = props.content.length;
      if (right - left <= 1) {
        if (props.direction === "end") {
          return {
            leading: props.content.slice(0, left) + props.symbol
          };
        } else {
          return {
            tailing: props.symbol + props.content.slice(right, end)
          };
        }
      }
      const middle = Math.round((left + right) / 2);
      if (props.direction === "end") {
        container.innerText = props.content.slice(0, middle) + props.symbol + actionText;
      } else {
        container.innerText = actionText + props.symbol + props.content.slice(middle, end);
      }
      if (container.offsetHeight <= maxHeight) {
        if (props.direction === "end") {
          return tailor(middle, right);
        } else {
          return tailor(left, middle);
        }
      } else {
        if (props.direction === "end") {
          return tailor(left, middle);
        } else {
          return tailor(middle, right);
        }
      }
    };
    const tailorMiddle = (leftPart, rightPart) => {
      const actionText = state.expanded ? props.collapseText : props.expandText;
      const end = props.content.length;
      if (leftPart[1] - leftPart[0] <= 1 && rightPart[1] - rightPart[0] <= 1) {
        return {
          leading: props.content.slice(0, leftPart[0]) + props.symbol,
          tailing: props.symbol + props.content.slice(rightPart[1], end)
        };
      }
      const leftPartMiddle = Math.floor((leftPart[0] + leftPart[1]) / 2);
      const rightPartMiddle = Math.ceil((rightPart[0] + rightPart[1]) / 2);
      container.innerText = props.content.slice(0, leftPartMiddle) + props.symbol + actionText + props.symbol + props.content.slice(rightPartMiddle, end);
      if (container.offsetHeight <= maxHeight) {
        return tailorMiddle([leftPartMiddle, leftPart[1]], [rightPart[0], rightPartMiddle]);
      } else {
        return tailorMiddle([leftPart[0], leftPartMiddle], [rightPartMiddle, rightPart[1]]);
      }
    };
    const pxToNumber = (value) => {
      if (!value)
        return 0;
      const match = value.match(/^\d*(\.\d*)?/);
      return match ? Number(match[0]) : 0;
    };
    const clickHandle = (type) => {
      if (type == 1) {
        state.expanded = true;
        emit("change", "expand");
      } else {
        state.expanded = false;
        emit("change", "collapse");
      }
    };
    const handleClick = () => {
      emit("click");
    };
    return { ...toRefs(state), root, ellipsis, classes, clickHandle, handleClick };
  }
});
const _hoisted_1 = { key: 0 };
const _hoisted_2 = { key: 1 };
const _hoisted_3 = { key: 2 };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    onClick: _cache[2] || (_cache[2] = (...args) => _ctx.handleClick && _ctx.handleClick(...args)),
    ref: "root"
  }, [
    !_ctx.exceeded ? (openBlock(), createElementBlock("view", _hoisted_1, toDisplayString(_ctx.content), 1)) : createCommentVNode("", true),
    _ctx.exceeded && !_ctx.expanded ? (openBlock(), createElementBlock("view", _hoisted_2, [
      createTextVNode(toDisplayString(_ctx.ellipsis.leading), 1),
      _ctx.expandText ? (openBlock(), createElementBlock("span", {
        key: 0,
        class: "nut-ellipsis-text",
        onClick: _cache[0] || (_cache[0] = withModifiers(($event) => _ctx.clickHandle(1), ["stop"]))
      }, toDisplayString(_ctx.expandText), 1)) : createCommentVNode("", true),
      createTextVNode(toDisplayString(_ctx.ellipsis.tailing), 1)
    ])) : createCommentVNode("", true),
    _ctx.exceeded && _ctx.expanded ? (openBlock(), createElementBlock("view", _hoisted_3, [
      createTextVNode(toDisplayString(_ctx.content) + " ", 1),
      _ctx.expandText ? (openBlock(), createElementBlock("span", {
        key: 0,
        class: "nut-ellipsis-text",
        onClick: _cache[1] || (_cache[1] = withModifiers(($event) => _ctx.clickHandle(2), ["stop"]))
      }, toDisplayString(_ctx.collapseText), 1)) : createCommentVNode("", true)
    ])) : createCommentVNode("", true)
  ], 2);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
