// @nutui/nutui/dist/packages/_es/Empty.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { toRefs, computed, openBlock, createElementBlock, createElementVNode, normalizeStyle, renderSlot, Fragment, createCommentVNode, toDisplayString } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create, translate } = createComponent("empty");
const defaultStatus = {
  empty: "https://static-ftcms.jd.com/p/files/61a9e3183985005b3958672b.png",
  error: "https://ftcms.jd.com/p/files/61a9e33ee7dcdbcc0ce62736.png",
  network: "https://static-ftcms.jd.com/p/files/61a9e31de7dcdbcc0ce62734.png"
};
const _sfc_main = create({
  props: {
    image: {
      type: String,
      default: "empty"
    },
    imageSize: {
      type: [Number, String],
      default: ""
    },
    description: {
      type: String,
      default: ""
    }
  },
  setup(props) {
    const { image, imageSize } = toRefs(props);
    const imgStyle = computed(() => {
      if (!imageSize.value) {
        return "";
      }
      if (typeof imageSize.value === "number") {
        return `width:${imageSize.value}px;height:${imageSize.value}px`;
      }
      return `width:${imageSize.value};height:${imageSize.value}`;
    });
    const isHttpUrl = image.value.startsWith("https://") || image.value.startsWith("http://") || image.value.startsWith("//");
    const imageUrl = isHttpUrl ? image.value : defaultStatus[image.value];
    return {
      imageUrl,
      imgStyle,
      translate
    };
  }
});
const _hoisted_1 = { class: "nut-empty" };
const _hoisted_2 = ["src"];
const _hoisted_3 = {
  key: 1,
  class: "nut-empty-description"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("view", _hoisted_1, [
    createElementVNode("view", {
      class: "nut-empty-image",
      style: normalizeStyle(_ctx.imgStyle)
    }, [
      _ctx.$slots.image ? renderSlot(_ctx.$slots, "image", { key: 0 }) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
        _ctx.imageUrl ? (openBlock(), createElementBlock("img", {
          key: 0,
          class: "img",
          src: _ctx.imageUrl
        }, null, 8, _hoisted_2)) : createCommentVNode("", true)
      ], 64))
    ], 4),
    _ctx.$slots.description ? renderSlot(_ctx.$slots, "description", { key: 0 }) : (openBlock(), createElementBlock("view", _hoisted_3, toDisplayString(_ctx.description || _ctx.translate("noData")), 1)),
    _ctx.$slots.default ? renderSlot(_ctx.$slots, "default", { key: 2 }) : createCommentVNode("", true)
  ]);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
