// @nutui/nutui/dist/packages/_es/SideNavBarItem.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { computed, openBlock, createElementBlock, withModifiers, createElementVNode, toDisplayString } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("sidenavbaritem");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    ikey: {
      type: String,
      default: ""
    }
  },
  emits: ["click"],
  setup: (props, context) => {
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const handleClick = () => {
      context.emit("click");
    };
    return {
      classes,
      handleClick
    };
  }
});
const _hoisted_1 = ["ikey"];
const _hoisted_2 = { class: "nut-sidenavbaritem__title" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("view", {
    class: "nut-sidenavbaritem",
    onClick: _cache[0] || (_cache[0] = withModifiers((...args) => _ctx.handleClick && _ctx.handleClick(...args), ["stop"])),
    ikey: _ctx.ikey
  }, [
    createElementVNode("span", _hoisted_2, toDisplayString(_ctx.title), 1)
  ], 8, _hoisted_1);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
