// @nutui/nutui/dist/packages/_es/Step.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { getCurrentInstance, inject, reactive, computed, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, createBlock, mergeProps, Fragment, toDisplayString, createCommentVNode, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { create, componentName } = createComponent("step");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    content: {
      type: String,
      default: ""
    },
    icon: {
      type: String,
      default: null
    },
    iconColor: {
      type: String,
      default: ""
    },
    size: {
      type: [String, Number],
      default: "12px"
    }
  },
  emits: ["click-step"],
  setup(props, { emit, slots }) {
    const { proxy } = getCurrentInstance();
    const parent = inject("parent");
    parent["relation"](proxy);
    const state = reactive({
      dot: parent.props.progressDot
    });
    const index2 = computed(() => parent.state.children.indexOf(proxy) + 1);
    const getCurrentStatus = () => {
      const activeIndex = index2.value;
      if (activeIndex < +parent.props.current)
        return "finish";
      return activeIndex === +parent.props.current ? "process" : "wait";
    };
    const status = computed(() => {
      return getCurrentStatus();
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true,
        [`${prefixCls}-${status.value}`]: true
      };
    });
    const handleClickStep = () => {
      parent["onEmit"](index2.value);
    };
    return {
      ...toRefs(state),
      index: index2,
      classes,
      handleClickStep
    };
  }
});
const _hoisted_1 = { class: "nut-step-head" };
const _hoisted_2 = /* @__PURE__ */ createElementVNode("view", { class: "nut-step-line" }, null, -1);
const _hoisted_3 = {
  key: 2,
  class: "nut-step-inner"
};
const _hoisted_4 = { class: "nut-step-main" };
const _hoisted_5 = { class: "nut-step-title" };
const _hoisted_6 = { key: 0 };
const _hoisted_7 = {
  key: 0,
  class: "nut-step-content"
};
const _hoisted_8 = ["innerHTML"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClickStep && _ctx.handleClickStep(...args))
  }, [
    createElementVNode("view", _hoisted_1, [
      _hoisted_2,
      createElementVNode("view", {
        class: normalizeClass(["nut-step-icon", [!_ctx.dot ? _ctx.icon ? "is-icon" : "is-text" : ""]])
      }, [
        _ctx.icon ? (openBlock(), createBlock(_component_nut_icon, mergeProps({
          key: 0,
          class: "nut-step-icon-inner"
        }, _ctx.$attrs, {
          color: _ctx.iconColor,
          name: _ctx.icon,
          size: _ctx.size
        }), null, 16, ["color", "name", "size"])) : _ctx.dot ? (openBlock(), createElementBlock(Fragment, { key: 1 }, [], 64)) : (openBlock(), createElementBlock("view", _hoisted_3, toDisplayString(_ctx.index), 1))
      ], 2)
    ]),
    createElementVNode("view", _hoisted_4, [
      createElementVNode("view", _hoisted_5, [
        !_ctx.$slots.title ? (openBlock(), createElementBlock("span", _hoisted_6, toDisplayString(_ctx.title), 1)) : createCommentVNode("", true),
        renderSlot(_ctx.$slots, "title")
      ]),
      _ctx.content || _ctx.$slots.content ? (openBlock(), createElementBlock("view", _hoisted_7, [
        !_ctx.$slots.content ? (openBlock(), createElementBlock("span", {
          key: 0,
          innerHTML: _ctx.content
        }, null, 8, _hoisted_8)) : createCommentVNode("", true),
        renderSlot(_ctx.$slots, "content")
      ])) : createCommentVNode("", true)
    ])
  ], 2);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
