// @nutui/nutui/dist/packages/_es/SubSideNavBar.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { reactive, computed, onMounted, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, withModifiers, toDisplayString, createVNode, normalizeStyle, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("subsidenavbar");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    ikey: {
      type: [String, Number],
      default: ""
    },
    open: {
      type: Boolean,
      default: true
    }
  },
  emits: ["title-click"],
  setup: (props, context) => {
    const state = reactive({
      direction: ""
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const style = computed(() => {
      return {
        height: !state.direction ? "auto" : "0px"
      };
    });
    const handleClick = () => {
      context.emit("title-click");
      state.direction = !state.direction ? "up" : "";
    };
    onMounted(() => {
      state.direction = props.open ? "" : "up";
    });
    return {
      ...toRefs(state),
      classes,
      style,
      handleClick
    };
  }
});
const _hoisted_1 = ["ikey"];
const _hoisted_2 = { class: "nut-subsidenavbar__title__text" };
const _hoisted_3 = { class: "nut-subsidenavbar__title__icon" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    ikey: _ctx.ikey
  }, [
    createElementVNode("view", {
      class: "nut-subsidenavbar__title",
      onClick: _cache[0] || (_cache[0] = withModifiers((...args) => _ctx.handleClick && _ctx.handleClick(...args), ["stop"]))
    }, [
      createElementVNode("span", _hoisted_2, toDisplayString(_ctx.title), 1),
      createElementVNode("span", _hoisted_3, [
        createVNode(_component_nut_icon, {
          name: "down-arrow",
          class: normalizeClass(_ctx.direction)
        }, null, 8, ["class"])
      ])
    ]),
    createElementVNode("view", {
      class: normalizeClass(["nut-subsidenavbar__list", !_ctx.direction ? "nutFadeIn" : "nutFadeOut"]),
      style: normalizeStyle(_ctx.style)
    }, [
      renderSlot(_ctx.$slots, "default")
    ], 6)
  ], 10, _hoisted_1);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
