// @nutui/nutui/dist/packages/_es/TrendArrow.js
/*!
* @nutui/nutui v3.3.8 Mon Jun 12 2023 11:24:18 GMT+0800 (中国标准时间)
* (c) 2022 @jdf2e.
* Released under the MIT License.
*/
import { reactive, computed, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, normalizeStyle, toDisplayString, createCommentVNode, renderSlot, createBlock } from "vue";
import { m as myFixed, c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
import "../locale/lang";
const { componentName, create } = createComponent("trendarrow");
const _sfc_main = create({
  props: {
    rate: {
      type: Number,
      default: 0
    },
    digits: {
      type: Number,
      default: 2
    },
    showSign: {
      type: Boolean,
      default: false
    },
    showZero: {
      type: Boolean,
      default: false
    },
    arrowLeft: {
      type: Boolean,
      default: false
    },
    syncTextColor: {
      type: Boolean,
      default: true
    },
    textColor: {
      type: String,
      default: "#333"
    },
    riseColor: {
      type: String,
      default: "#fa2c19"
    },
    dropColor: {
      type: String,
      default: "#64b578"
    },
    iconSize: {
      type: String,
      default: "12px"
    },
    upIconName: {
      type: String,
      default: "triangle-up"
    },
    downIconName: {
      type: String,
      default: "triangle-down"
    }
  },
  setup(props) {
    const state = reactive({
      rateTrend: props.rate > 0 ? true : false
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const calcRate = computed(() => {
      const { rate, digits, showSign, showZero } = props;
      state.rateTrend = rate > 0 ? true : false;
      const absRate = Math.abs(rate);
      if (!showZero && rate === 0) {
        return "--";
      }
      let resultRate = `${showSign && rate !== 0 ? state.rateTrend ? "+" : "-" : ""}${myFixed(
        Number(absRate),
        digits
      )}%`;
      return resultRate;
    });
    const calcStyle = computed(() => {
      const { dropColor, riseColor, syncTextColor, textColor, rate } = props;
      let style = {
        color: rate === 0 ? textColor : syncTextColor ? state.rateTrend ? riseColor : dropColor : textColor
      };
      return style;
    });
    const calcIconProps = computed(() => {
      const { dropColor, riseColor, iconSize, upIconName, downIconName } = props;
      let iconProps = {
        name: state.rateTrend ? upIconName : downIconName,
        color: state.rateTrend ? riseColor : dropColor,
        size: iconSize
      };
      return iconProps;
    });
    return { ...toRefs(state), classes, calcRate, calcStyle, calcIconProps };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    !_ctx.arrowLeft ? (openBlock(), createElementBlock("span", {
      key: 0,
      class: "nut-trendarrow-icon-before nut-trendarrow-rate",
      style: normalizeStyle(_ctx.calcStyle)
    }, toDisplayString(_ctx.calcRate), 5)) : createCommentVNode("", true),
    renderSlot(_ctx.$slots, "default", {}, () => [
      Number(_ctx.rate) !== 0 ? (openBlock(), createBlock(_component_nut_icon, {
        key: 0,
        size: _ctx.calcIconProps.size,
        name: _ctx.calcIconProps.name,
        color: _ctx.calcIconProps.color
      }, null, 8, ["size", "name", "color"])) : createCommentVNode("", true)
    ]),
    _ctx.arrowLeft ? (openBlock(), createElementBlock("span", {
      key: 1,
      class: "nut-trendarrow-icon-after nut-trendarrow-rate",
      style: normalizeStyle(_ctx.calcStyle)
    }, toDisplayString(_ctx.calcRate), 5)) : createCommentVNode("", true)
  ], 2);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
