// @antv/f2/lib/components/axis/polar/polar-y.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _jsx = require("../../../jsx");
var _f2Graphic = require("@antv/f2-graphic");
var _default = function _default(props) {
  var ticks = props.ticks,
    coord = props.coord,
    style = props.style,
    gridType = props.grid;
  var center = coord.center;
  var grid = style.grid,
    tickLine = style.tickLine,
    line = style.line,
    labelOffset = style.labelOffset,
    label = style.label;
  return (0, _jsx.jsx)("group", null, grid ? ticks.map(function (tick) {
    var points = tick.points,
      gridStyle = tick.gridStyle,
      gridPoints = tick.gridPoints;
    var end = points[points.length - 1];
    if (gridType !== 'line') {
      return (0, _jsx.jsx)("arc", {
        attrs: (0, _objectSpread2.default)((0, _objectSpread2.default)({
          x: center.x,
          y: center.y,
          r: _f2Graphic.Vector2.length([end.x - center.x, end.y - center.y])
        }, grid), gridStyle)
      });
    }
    return (0, _jsx.jsx)("polyline", {
      attrs: (0, _objectSpread2.default)((0, _objectSpread2.default)({
        points: gridPoints
      }, grid), gridStyle)
    });
  }) : null, tickLine && tickLine.length ? ticks.map(function (tick) {
    var points = tick.points;
    var end = points[points.length - 1];
    return (0, _jsx.jsx)("line", {
      attrs: (0, _objectSpread2.default)({
        x1: end.x,
        y1: end.y,
        x2: end.x - tickLine.length,
        y2: end.y
      }, tickLine)
    });
  }) : null, line ? (0, _jsx.jsx)("line", {
    attrs: (0, _objectSpread2.default)({
      x1: ticks[0].points[0].x,
      y1: ticks[0].points[0].y,
      x2: ticks[ticks.length - 1].points[0].x,
      y2: ticks[ticks.length - 1].points[0].y
    }, line)
  }) : null, label ? ticks.map(function (tick) {
    var points = tick.points,
      text = tick.text,
      labelStyle = tick.labelStyle;
    var end = points[points.length - 1];
    return (0, _jsx.jsx)("text", {
      attrs: (0, _objectSpread2.default)((0, _objectSpread2.default)({
        x: end.x - labelOffset,
        y: end.y,
        text: text,
        textAlign: 'right',
        textBaseline: 'middle'
      }, label), labelStyle)
    });
  }) : null);
};
exports.default = _default;
