// @antv/f2/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Children: true,
  Component: true,
  Timeline: true,
  Canvas: true,
  Chart: true,
  createElement: true,
  jsx: true,
  render: true,
  Fragment: true,
  renderShape: true,
  createRef: true
};
Object.defineProperty(exports, "Canvas", {
  enumerable: true,
  get: function get() {
    return _canvas.default;
  }
});
Object.defineProperty(exports, "Chart", {
  enumerable: true,
  get: function get() {
    return _chart.default;
  }
});
Object.defineProperty(exports, "Children", {
  enumerable: true,
  get: function get() {
    return _children.default;
  }
});
Object.defineProperty(exports, "Component", {
  enumerable: true,
  get: function get() {
    return _component.default;
  }
});
Object.defineProperty(exports, "Fragment", {
  enumerable: true,
  get: function get() {
    return _jsx.Fragment;
  }
});
Object.defineProperty(exports, "Timeline", {
  enumerable: true,
  get: function get() {
    return _timeline.default;
  }
});
Object.defineProperty(exports, "createElement", {
  enumerable: true,
  get: function get() {
    return _jsx.jsx;
  }
});
Object.defineProperty(exports, "createRef", {
  enumerable: true,
  get: function get() {
    return _createRef.default;
  }
});
Object.defineProperty(exports, "jsx", {
  enumerable: true,
  get: function get() {
    return _jsx.jsx;
  }
});
Object.defineProperty(exports, "render", {
  enumerable: true,
  get: function get() {
    return _jsx.render;
  }
});
Object.defineProperty(exports, "renderShape", {
  enumerable: true,
  get: function get() {
    return _diff.renderShape;
  }
});
var _children = _interopRequireDefault(require("./children"));
var _component = _interopRequireDefault(require("./base/component"));
var _timeline = _interopRequireDefault(require("./timeline"));
var _canvas = _interopRequireDefault(require("./canvas"));
var _chart = _interopRequireDefault(require("./chart"));
var _jsx = require("./jsx");
var _diff = require("./base/diff");
var _components = require("./components");
Object.keys(_components).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _components[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _components[key];
    }
  });
});
var _createRef = _interopRequireDefault(require("./createRef"));
