// @antv/f2/es/components/area/withArea.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _get from "@babel/runtime/helpers/esm/get";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import withLine from '../line/withLine';
export default (function (View) {
  return /*#__PURE__*/function (_withLine) {
    _inherits(Area, _withLine);
    var _super = _createSuper(Area);
    function Area() {
      _classCallCheck(this, Area);
      return _super.apply(this, arguments);
    }
    _createClass(Area, [{
      key: "getDefaultCfg",
      value: function getDefaultCfg() {
        return {
          geomType: 'area',
          // 面积图默认设为从0开始
          startOnZero: true,
          // 点需要排序
          sortable: true
        };
      }
    }, {
      key: "mapping",
      value: function mapping() {
        var records = _get(_getPrototypeOf(Area.prototype), "mapping", this).call(this);
        // 坐标轴 y0
        var y0 = this.getY0Value();
        var props = this.props,
          defaultStartOnZero = this.startOnZero;
        var coord = props.coord,
          _props$startOnZero = props.startOnZero,
          startOnZero = _props$startOnZero === void 0 ? defaultStartOnZero : _props$startOnZero;
        var baseY = coord.y[0];
        if (startOnZero) {
          // 零点映射到绝对坐标
          var originCoord = coord.convertPoint({
            x: 0,
            y: y0
          });
          baseY = originCoord.y;
        }
        for (var i = 0, len = records.length; i < len; i++) {
          var record = records[i];
          var children = record.children;
          for (var j = 0, _len = children.length; j < _len; j++) {
            var child = children[j];
            var points = child.points,
              bottomPoints = child.bottomPoints;
            if (bottomPoints && bottomPoints.length) {
              bottomPoints.reverse();
              child.points = points.concat(bottomPoints);
            } else {
              points.push({
                x: points[points.length - 1].x,
                y: baseY
              });
              points.push({
                x: points[0].x,
                y: baseY
              });
            }
          }
        }
        return records;
      }
    }]);
    return Area;
  }(withLine(View));
});
