// @antv/f2/es/components/point/pointView.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { deepMix, isNil } from '@antv/util';
import { jsx } from '../../jsx';
export default (function (props) {
  var records = props.records,
    animation = props.animation,
    clip = props.clip;
  return jsx("group", {
    attrs: {
      clip: clip
    }
  }, records.map(function (record) {
    var key = record.key,
      children = record.children;
    return jsx("group", {
      key: key
    }, children.map(function (item) {
      var x = item.x,
        y = item.y,
        size = item.size,
        color = item.color,
        shapeName = item.shapeName,
        shape = item.shape;
      if (isNaN(x) || isNaN(y)) {
        return null;
      }
      if (shapeName === 'rect') {
        var rectSize = isNil(size) ? shape.size : size;
        return jsx("rect", {
          attrs: _objectSpread(_objectSpread({
            x: x - rectSize,
            y: y - rectSize,
            fill: color,
            stroke: color
          }, shape), {}, {
            width: rectSize * 2,
            height: rectSize * 2
          }),
          animation: deepMix({
            appear: {
              easing: 'linear',
              duration: 450
            },
            update: {
              easing: 'linear',
              duration: 450,
              property: ['x', 'y', 'width', 'height', 'fill']
            }
          }, animation)
        });
      }
      return jsx("circle", {
        attrs: _objectSpread(_objectSpread({
          x: x,
          y: y,
          fill: shapeName === 'circle' ? color : null,
          stroke: shapeName === 'hollowCircle' ? color : null
        }, shape), {}, {
          r: isNil(size) ? shape.size : size
        }),
        animation: deepMix({
          appear: {
            easing: 'linear',
            duration: 450
          },
          update: {
            easing: 'linear',
            duration: 450,
            property: ['x', 'y', 'r', 'fill']
          }
        }, animation)
      });
    }));
  }));
});
