// @antv/f2/lib/components/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ArcGuide", {
  enumerable: true,
  get: function get() {
    return _guide.ArcGuide;
  }
});
Object.defineProperty(exports, "Area", {
  enumerable: true,
  get: function get() {
    return _area.default;
  }
});
Object.defineProperty(exports, "AreaView", {
  enumerable: true,
  get: function get() {
    return _area.AreaView;
  }
});
Object.defineProperty(exports, "Axis", {
  enumerable: true,
  get: function get() {
    return _axis.default;
  }
});
Object.defineProperty(exports, "AxisView", {
  enumerable: true,
  get: function get() {
    return _axis.AxisView;
  }
});
Object.defineProperty(exports, "Gauge", {
  enumerable: true,
  get: function get() {
    return _gauge.default;
  }
});
Object.defineProperty(exports, "GaugeView", {
  enumerable: true,
  get: function get() {
    return _gauge.GaugeView;
  }
});
Object.defineProperty(exports, "Geometry", {
  enumerable: true,
  get: function get() {
    return _geometry.default;
  }
});
Object.defineProperty(exports, "Guide", {
  enumerable: true,
  get: function get() {
    return _guide.default;
  }
});
Object.defineProperty(exports, "ImageGuide", {
  enumerable: true,
  get: function get() {
    return _guide.ImageGuide;
  }
});
Object.defineProperty(exports, "Interval", {
  enumerable: true,
  get: function get() {
    return _interval.default;
  }
});
Object.defineProperty(exports, "IntervalView", {
  enumerable: true,
  get: function get() {
    return _interval.IntervalView;
  }
});
Object.defineProperty(exports, "Legend", {
  enumerable: true,
  get: function get() {
    return _legend.default;
  }
});
Object.defineProperty(exports, "LegendView", {
  enumerable: true,
  get: function get() {
    return _legend.LegendView;
  }
});
Object.defineProperty(exports, "Line", {
  enumerable: true,
  get: function get() {
    return _line.default;
  }
});
Object.defineProperty(exports, "LineGuide", {
  enumerable: true,
  get: function get() {
    return _guide.LineGuide;
  }
});
Object.defineProperty(exports, "LineView", {
  enumerable: true,
  get: function get() {
    return _line.LineView;
  }
});
Object.defineProperty(exports, "PieLabel", {
  enumerable: true,
  get: function get() {
    return _pieLabel.default;
  }
});
Object.defineProperty(exports, "PieLabelView", {
  enumerable: true,
  get: function get() {
    return _pieLabel.PieLabelView;
  }
});
Object.defineProperty(exports, "Point", {
  enumerable: true,
  get: function get() {
    return _point.default;
  }
});
Object.defineProperty(exports, "PointGuide", {
  enumerable: true,
  get: function get() {
    return _guide.PointGuide;
  }
});
Object.defineProperty(exports, "PointView", {
  enumerable: true,
  get: function get() {
    return _point.PointView;
  }
});
Object.defineProperty(exports, "RectGuide", {
  enumerable: true,
  get: function get() {
    return _guide.RectGuide;
  }
});
Object.defineProperty(exports, "ScrollBar", {
  enumerable: true,
  get: function get() {
    return _scrollBar.default;
  }
});
Object.defineProperty(exports, "ScrollBarView", {
  enumerable: true,
  get: function get() {
    return _scrollBar.ScrollBarView;
  }
});
Object.defineProperty(exports, "Sunburst", {
  enumerable: true,
  get: function get() {
    return _sunburst.default;
  }
});
Object.defineProperty(exports, "SunburstView", {
  enumerable: true,
  get: function get() {
    return _sunburst.SunburstView;
  }
});
Object.defineProperty(exports, "TagGuide", {
  enumerable: true,
  get: function get() {
    return _guide.TagGuide;
  }
});
Object.defineProperty(exports, "TextGuide", {
  enumerable: true,
  get: function get() {
    return _guide.TextGuide;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _tooltip.default;
  }
});
Object.defineProperty(exports, "TooltipView", {
  enumerable: true,
  get: function get() {
    return _tooltip.TooltipView;
  }
});
Object.defineProperty(exports, "Treemap", {
  enumerable: true,
  get: function get() {
    return _treemap.default;
  }
});
Object.defineProperty(exports, "TreemapView", {
  enumerable: true,
  get: function get() {
    return _treemap.TreemapView;
  }
});
Object.defineProperty(exports, "Zoom", {
  enumerable: true,
  get: function get() {
    return _zoom.default;
  }
});
Object.defineProperty(exports, "withArea", {
  enumerable: true,
  get: function get() {
    return _area.withArea;
  }
});
Object.defineProperty(exports, "withAxis", {
  enumerable: true,
  get: function get() {
    return _axis.withAxis;
  }
});
Object.defineProperty(exports, "withGauge", {
  enumerable: true,
  get: function get() {
    return _gauge.withGauge;
  }
});
Object.defineProperty(exports, "withGuide", {
  enumerable: true,
  get: function get() {
    return _guide.withGuide;
  }
});
Object.defineProperty(exports, "withInterval", {
  enumerable: true,
  get: function get() {
    return _interval.withInterval;
  }
});
Object.defineProperty(exports, "withLegend", {
  enumerable: true,
  get: function get() {
    return _legend.withLegend;
  }
});
Object.defineProperty(exports, "withLine", {
  enumerable: true,
  get: function get() {
    return _line.withLine;
  }
});
Object.defineProperty(exports, "withPieLabel", {
  enumerable: true,
  get: function get() {
    return _pieLabel.withPieLabel;
  }
});
Object.defineProperty(exports, "withPoint", {
  enumerable: true,
  get: function get() {
    return _point.withPoint;
  }
});
Object.defineProperty(exports, "withScrollBar", {
  enumerable: true,
  get: function get() {
    return _scrollBar.withScrollBar;
  }
});
Object.defineProperty(exports, "withSunburst", {
  enumerable: true,
  get: function get() {
    return _sunburst.withSunburst;
  }
});
Object.defineProperty(exports, "withTooltip", {
  enumerable: true,
  get: function get() {
    return _tooltip.withTooltip;
  }
});
Object.defineProperty(exports, "withTreemap", {
  enumerable: true,
  get: function get() {
    return _treemap.withTreemap;
  }
});
var _geometry = _interopRequireDefault(require("./geometry"));
var _line = _interopRequireWildcard(require("./line"));
var _area = _interopRequireWildcard(require("./area"));
var _interval = _interopRequireWildcard(require("./interval"));
var _point = _interopRequireWildcard(require("./point"));
var _axis = _interopRequireWildcard(require("./axis"));
var _legend = _interopRequireWildcard(require("./legend"));
var _guide = _interopRequireWildcard(require("./guide"));
var _tooltip = _interopRequireWildcard(require("./tooltip"));
var _treemap = _interopRequireWildcard(require("./treemap"));
var _sunburst = _interopRequireWildcard(require("./sunburst"));
var _pieLabel = _interopRequireWildcard(require("./pieLabel"));
var _gauge = _interopRequireWildcard(require("./gauge"));
var _zoom = _interopRequireDefault(require("./zoom"));
var _scrollBar = _interopRequireWildcard(require("./scrollBar"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
