// @antv/f6-hammerjs/lib/inputjs/input-consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PROPS_CLIENT_XY = exports.PROPS_XY = exports.DIRECTION_ALL = exports.DIRECTION_VERTICAL = exports.DIRECTION_HORIZONTAL = exports.DIRECTION_DOWN = exports.DIRECTION_UP = exports.DIRECTION_RIGHT = exports.DIRECTION_LEFT = exports.DIRECTION_NONE = exports.INPUT_CANCEL = exports.INPUT_END = exports.INPUT_MOVE = exports.INPUT_START = exports.COMPUTE_INTERVAL = exports.INPUT_TYPE_TOUCH = exports.INPUT_TYPE_PEN = exports.INPUT_TYPE_MOUSE = exports.INPUT_TYPE_KINECT = void 0;
const INPUT_TYPE_TOUCH = 'touch';
exports.INPUT_TYPE_TOUCH = INPUT_TYPE_TOUCH;
const INPUT_TYPE_PEN = 'pen';
exports.INPUT_TYPE_PEN = INPUT_TYPE_PEN;
const INPUT_TYPE_MOUSE = 'mouse';
exports.INPUT_TYPE_MOUSE = INPUT_TYPE_MOUSE;
const INPUT_TYPE_KINECT = 'kinect';
exports.INPUT_TYPE_KINECT = INPUT_TYPE_KINECT;
const COMPUTE_INTERVAL = 25;
exports.COMPUTE_INTERVAL = COMPUTE_INTERVAL;
const INPUT_START = 1;
exports.INPUT_START = INPUT_START;
const INPUT_MOVE = 2;
exports.INPUT_MOVE = INPUT_MOVE;
const INPUT_END = 4;
exports.INPUT_END = INPUT_END;
const INPUT_CANCEL = 8;
exports.INPUT_CANCEL = INPUT_CANCEL;
const DIRECTION_NONE = 1;
exports.DIRECTION_NONE = DIRECTION_NONE;
const DIRECTION_LEFT = 2;
exports.DIRECTION_LEFT = DIRECTION_LEFT;
const DIRECTION_RIGHT = 4;
exports.DIRECTION_RIGHT = DIRECTION_RIGHT;
const DIRECTION_UP = 8;
exports.DIRECTION_UP = DIRECTION_UP;
const DIRECTION_DOWN = 16;
exports.DIRECTION_DOWN = DIRECTION_DOWN;
const DIRECTION_HORIZONTAL = DIRECTION_LEFT | DIRECTION_RIGHT;
exports.DIRECTION_HORIZONTAL = DIRECTION_HORIZONTAL;
const DIRECTION_VERTICAL = DIRECTION_UP | DIRECTION_DOWN;
exports.DIRECTION_VERTICAL = DIRECTION_VERTICAL;
const DIRECTION_ALL = DIRECTION_HORIZONTAL | DIRECTION_VERTICAL;
exports.DIRECTION_ALL = DIRECTION_ALL;
const PROPS_XY = ['x', 'y'];
exports.PROPS_XY = PROPS_XY;
const PROPS_CLIENT_XY = ['clientX', 'clientY'];
exports.PROPS_CLIENT_XY = PROPS_CLIENT_XY;
