// @antv/f6-hammerjs/lib/inputjs/simple-clone-input-data.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = simpleCloneInputData;

var _utilsConsts = require("../utils/utils-consts");

var _getCenter = _interopRequireDefault(require("./get-center"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @private
 * create a simple clone from the input used for storage of firstInput and firstMultiple
 * @param {Object} input
 * @returns {Object} clonedInputData
 */
function simpleCloneInputData(input) {
  // make a simple copy of the pointers because we will get a reference if we don't
  // we only need clientXY for the calculations
  let pointers = [];
  let i = 0;

  while (i < input.pointers.length) {
    pointers[i] = {
      clientX: (0, _utilsConsts.round)(input.pointers[i].clientX),
      clientY: (0, _utilsConsts.round)(input.pointers[i].clientY)
    };
    i++;
  }

  return {
    timeStamp: (0, _utilsConsts.now)(),
    pointers,
    center: (0, _getCenter.default)(pointers),
    deltaX: input.deltaX,
    deltaY: input.deltaY
  };
}
