// @antv/f2/es/components/scrollBar/horizontal.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { jsx } from '../../jsx';
export default (function (props, context) {
  var coord = props.coord,
    range = props.range,
    position = props.position,
    layout = props.layout;
  var left = coord.left,
    width = coord.width;
  var top = layout.top,
    height = layout.height;
  var _ref = (range === null || range === void 0 ? void 0 : range.x) || (range === null || range === void 0 ? void 0 : range.y),
    _ref2 = _slicedToArray(_ref, 2),
    start = _ref2[0],
    end = _ref2[1];
  var barLeft = width * start;
  var barWidth = width * (end - start);
  return jsx("group", {
    style: {
      left: left,
      top: position === 'top' ? top - context.px2hd('8px') : top + height
    }
  }, jsx("line", {
    style: {
      position: 'absolute',
      left: 0,
      width: width,
      height: 0
    },
    attrs: {
      stroke: 'rgba(202, 215, 239, .2)',
      lineCap: 'round',
      lineWidth: '8px'
    }
  }), jsx("line", {
    style: {
      position: 'absolute',
      left: barLeft,
      width: barWidth,
      height: 0
    },
    attrs: {
      stroke: 'rgba(202, 215, 239, .5)',
      lineCap: 'round',
      lineWidth: '8px'
    }
  }));
});
