// @antv/f2/es/components/scrollBar/vertical.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { jsx } from '../../jsx';
export default (function (props, context) {
  var coord = props.coord,
    range = props.range,
    position = props.position,
    layout = props.layout;
  var top = coord.top,
    height = coord.height;
  var left = layout.left,
    width = layout.width;
  var _ref = (range === null || range === void 0 ? void 0 : range.y) || (range === null || range === void 0 ? void 0 : range.x),
    _ref2 = _slicedToArray(_ref, 2),
    start = _ref2[0],
    end = _ref2[1];
  var barTop = height * start;
  var barHeight = height * (end - start);
  return jsx("group", {
    style: {
      top: top,
      left: position === 'left' ? left - context.px2hd('8px') : left + width
    }
  }, jsx("line", {
    style: {
      position: 'absolute',
      top: 0,
      left: 0,
      width: 0,
      height: height
    },
    attrs: {
      stroke: 'rgba(202, 215, 239, .2)',
      lineCap: 'round',
      lineWidth: '8px'
    }
  }), jsx("line", {
    style: {
      position: 'absolute',
      top: barTop,
      width: 0,
      height: barHeight
    },
    attrs: {
      stroke: 'rgba(202, 215, 239, .5)',
      lineCap: 'round',
      lineWidth: '8px'
    }
  }));
});
