// @antv/f2/lib/controller/coord.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _coord = require("../coord");
var _util = require("@antv/util");
var coordMap = {
  rect: _coord.Rect,
  polar: _coord.Polar
};
var coordController = /*#__PURE__*/function () {
  function coordController() {
    (0, _classCallCheck2.default)(this, coordController);
  }
  (0, _createClass2.default)(coordController, [{
    key: "getOption",
    value: function getOption(cfg) {
      if ((0, _util.isString)(cfg)) {
        return {
          type: coordMap[cfg] || _coord.Rect
        };
      }
      if ((0, _util.isFunction)(cfg)) {
        return {
          type: cfg
        };
      }
      var _ref = cfg || {},
        type = _ref.type;
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, cfg), {}, {
        // 默认直角坐标系
        type: (0, _util.isFunction)(type) ? type : coordMap[type] || _coord.Rect
      });
    }
  }, {
    key: "create",
    value: function create(cfg, layout) {
      var option = this.getOption(cfg);
      var type = option.type;
      var coord = new type((0, _objectSpread2.default)((0, _objectSpread2.default)({}, option), layout));
      this.coord = coord;
      return coord;
    }
  }, {
    key: "updateLayout",
    value: function updateLayout(layout) {
      this.coord.update(layout);
    }
  }, {
    key: "update",
    value: function update() {}
  }]);
  return coordController;
}();
var _default = coordController;
exports.default = _default;
