// @antv/f2/lib/controller/layout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _layout = _interopRequireDefault(require("../base/layout"));
var LayoutController = /*#__PURE__*/function () {
  function LayoutController() {
    (0, _classCallCheck2.default)(this, LayoutController);
  }
  (0, _createClass2.default)(LayoutController, [{
    key: "getRectRange",
    value: function getRectRange(style) {
      var left = style.left,
        top = style.top,
        width = style.width,
        height = style.height,
        padding = style.padding;
      var _padding = (0, _slicedToArray2.default)(padding, 4),
        paddingTop = _padding[0],
        paddingRight = _padding[1],
        paddingBottom = _padding[2],
        paddingLeft = _padding[3];
      return {
        left: left + paddingLeft,
        top: top + paddingTop,
        width: width - paddingLeft - paddingRight,
        height: height - paddingTop - paddingBottom
      };
    }
  }, {
    key: "create",
    value: function create(style) {
      var rectRange = this.getRectRange(style);
      var layout = new _layout.default(rectRange);
      this.layout = layout;
      return layout;
    }
  }, {
    key: "update",
    value: function update(style) {
      var rectRange = this.getRectRange(style);
      var layout = this.layout;
      layout.update(rectRange);
      return layout;
    }
  }]);
  return LayoutController;
}();
var _default = LayoutController;
exports.default = _default;
