// @antv/f2-graphic/lib/engine/shape/shape.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));
var _util = require("@antv/util");
var _element = _interopRequireDefault(require("../element"));
var Shape = /*#__PURE__*/function (_Element) {
  (0, _inherits2.default)(Shape, _Element);
  var _super = (0, _createSuper2.default)(Shape);
  function Shape() {
    (0, _classCallCheck2.default)(this, Shape);
    return _super.apply(this, arguments);
  }
  (0, _createClass2.default)(Shape, [{
    key: "_initProperties",
    value: /* eslint-enable */
    function _initProperties() {
      this._attrs = (0, _objectSpread2.default)((0, _objectSpread2.default)({}, this._attrs), {}, {
        zIndex: 0,
        visible: true,
        destroyed: false,
        isShape: true,
        attrs: {}
      });
    }
  }, {
    key: "getType",
    value: function getType() {
      return this._attrs.type;
    }
  }, {
    key: "drawInner",
    value: function drawInner(context) {
      var attrs = this.get('attrs');
      this.createPath(context);
      var originOpacity = context.globalAlpha;
      if (this.hasFill()) {
        var fillOpacity = attrs.fillOpacity;
        if (!(0, _util.isNil)(fillOpacity) && fillOpacity !== 1) {
          context.globalAlpha = fillOpacity;
          context.fill();
          context.globalAlpha = originOpacity;
        } else {
          context.fill();
        }
      }
      if (this.hasStroke()) {
        var lineWidth = attrs.lineWidth;
        if (lineWidth > 0) {
          var strokeOpacity = attrs.strokeOpacity;
          if (!(0, _util.isNil)(strokeOpacity) && strokeOpacity !== 1) {
            context.globalAlpha = strokeOpacity;
          }
          context.stroke();
        }
      }
    }
  }, {
    key: "getBBox",
    value: function getBBox() {
      var bbox = this._attrs.bbox;
      if (!bbox) {
        bbox = this.calculateBox();
        if (bbox) {
          bbox.x = bbox.minX;
          bbox.y = bbox.minY;
          bbox.width = bbox.maxX - bbox.minX;
          bbox.height = bbox.maxY - bbox.minY;
        }
        this._attrs.bbox = bbox;
      }
      return bbox;
    }
  }, {
    key: "calculateBox",
    value: function calculateBox() {
      return null;
    }
  }, {
    key: "createPath",
    value: function createPath(_context) {}
  }]);
  return Shape;
}(_element.default);
var _default = Shape;
exports.default = _default;
