// @antv/f2/es/components/guide/views/Text.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import { jsx } from '../../../jsx';
import { deepMix } from '@antv/util';
export default (function (props, context) {
  var _props$theme = props.theme,
    theme = _props$theme === void 0 ? {} : _props$theme;
  var _deepMix = deepMix(_objectSpread({}, theme.text), props),
    points = _deepMix.points,
    style = _deepMix.style,
    offsetX = _deepMix.offsetX,
    offsetY = _deepMix.offsetY,
    content = _deepMix.content,
    animation = _deepMix.animation;
  var _ref = points[0] || {},
    x = _ref.x,
    y = _ref.y;
  var offsetXNum = context.px2hd(offsetX);
  var offsetYNum = context.px2hd(offsetY);
  var posX = x + (offsetXNum || 0);
  var posY = y + (offsetYNum || 0);
  return jsx("text", {
    attrs: _objectSpread({
      text: content,
      x: posX,
      y: posY
    }, style),
    animation: deepMix({
      update: {
        easing: 'linear',
        duration: 450,
        property: ['x', 'y']
      }
    }, animation)
  });
});
